vernac/comProgramFixpoint.ml

```diff
@@ -232,18 +232,17 @@ let build_wellfounded pm (recname,pl,bl,arityc,body) poly ?typing_flags ?user_wa
     sigma, mkApp (h_a_term, [| argtyp ; wf_rel ; h_e_term; prop |])
   in
   let sigma, def = Typing.solve_evars env sigma def in
-  let sigma = Evd.collapse_sort_variables sigma in
   let sigma = Evarutil.nf_evar_map sigma in
   let def = mkApp (def, [|intern_body_lam|]) in
   let binders_rel = nf_evar_context sigma binders_rel in
   let binders = nf_evar_context sigma binders in
   let top_arity = Evarutil.nf_evar sigma top_arity in
+  let make = Evarutil.nf_evar sigma make in
   let hook, recname, typ =
     if List.length binders_rel > 1 then
       let name = add_suffix recname "_func" in
-      (* XXX: Mutating the evar_map in the hook! *)
-      (* XXX: Likely the sigma is out of date when the hook is called .... *)
-      let hook sigma { Declare.Hook.S.dref; _ } =
+      let hook { Declare.Hook.S.dref; uctx; _ } =
+        let sigma = Evd.from_ctx uctx in
```

Comment:
I wonder why we simply don't pass an evarmap in the hook argument, instead of reconstructing it from uctx. I think I tried something like that at some point because Equations hacks around this precise limitation, but since I have no idea about the expected invariants it is a bit hard to reason about.

Reply:
The idea I guess is that a hook is applied on declarations whose body and type are evar-free. But if a full sigma is needed for `Equations` why not to pass a sigma.

Reply:
@ppedrot:

Maybe is your suggestion to pass a sigma good. Somehow, it depends on what are the projects with `Program Fixpoint`? Are there projects on the Nantes side to bind it on top of the proof engine? Or should we set such project?

At least, if we keep the obligation mechanism as it is, it has bugs and we have to do something. For instance, the following variation on `bug_4725.v` breaks:
```coq
Program Fixpoint nubV'' (A:_ * _) V eqV equivV `{eqDecV : @EqDec V eqV equivV} (l : list V)
        {  measure (@length V l) lt } :=
    match l with
      | nil => nil
      | x::xs => x :: @nubV'' A V eqV equivV eqDecV (removeV x xs) _
    end.
Next Obligation. exact nat. Defined.
Next Obligation. exact nat. Defined.
Next Obligation. apply remove_le. Defined.
(* Anomaly "in econstr: grounding a non evar-free term" *)
```
and the reason is that `remove_obligations` collects evars into obligations but does not define the evars with the constants associated to the obligations. So, in the hook of `build_founded`, we have subterms (namely here the two `_` in `A : _ * _`) that are still evars and which are still undefined in the evar_map, and we have no simple way to instantiate them with the expected obligation constant name.

We could imagine different ways to instantiate them:
- e.g. to let `declare.ml` do the update of the evar_map when the obligations are defined and let the hook take the updated evar_map;
- or to manually recover the instances of the evars from the data of `retrieve_obligations` in the code of the hook in `ComProgramFixpoint`.

But before implementing a direction, knowing that it would in all cases means agreeing on a conventional way to do things, I was going to know the projects at Nantes on this matter.